content/BPIP-3.md

```diff
@@ -146,28 +149,113 @@ The following method is added.
+
+Given that now the buyer does not directly interact with the protocol, this opens the question of how to ensure protocol exchange guarantees (i.e. possibility of getting funds back if something goes wrong). The proposed solution is that now the seller needs to provide enough funds to cover both the item price and the seller deposit. This immediately ensures buyer protection but increases a seller's capital requirement a bit. However, this drawback is not too big, since as soon as the marketplace exchange happens, the seller receives the funds, which can be directly used to cover the item price in subsequent sales. Moreover, as soon as some exchanges are finalized and funds are released back into the seller's pool, those funds can be automatically used to cover future sales.
+
+This proposal might raise some efficiency concerns about the solution efficiency since minting ERC721 is a relatively costly operation. ERC1155 was considered as an alternative, but it turned out that with it it's impossible to maintain a nice exchange id and token id equivalence. To overcome the problem of ERC721 inefficiency, we propose another solution. Since premint always assigns token ownership to the voucher contract owner, there is no need to explicitly store its address during the premint. It's enough that only correct transfer events are emitted, while general consistency is achieved by overriding ERC721 `ownerOf`, `transferFrom` and `safeTransferFrom` methods to correctly manage all tokens that were preminted.
 
 ## Backward compatibility
 This specification does not break backward compatibility.
 
+Still it is important that when the upgrade is done, voucher contracts are upgraded before the protocol is upgraded.
+
 ## Implementation
 ### BosonVoucher
-* Support the tracking of reserved ranges so that exchange id and token id can remain the same.
-* Emit events that look as if vouchers have been minted, but don't store the owner address to conserve the bulk of an actual minting's gas. 
-* In the ownerOf method, if the token id is in a reserved range and has no stored owner, report the contract owner (the seller) as the owner of the voucher. 
-* Add a hook transfer hook that will detect if the voucher is being transferred to the first real owner, and if so, call the commitToPremintedOffer protocol method, which does not require payment.
+
+* Range reservation
+  * Must be done before preminting can happen.
+  * Ranges must be strictly increasing, i.e. start of a new range must be greater than the current highest range.
+  * Store information about the associated offer id, range start, and range length.
```

Comment:
And the address `_to` of who the owner is (i.e. contract or contract owner)

Reply:
Added